server/routerlicious/packages/services/src/mongoExceptionRetryRules/mongoError/index.ts

```diff
@@ -0,0 +1,183 @@
+// Should not retry but relays on reconnect.
+class NoPrimaryInReplicasetRule implements IMongoExceptionRetryRule {
+    private static readonly message = "no primary found in replicaset or invalid replica set name";
+    match(error: any): boolean {
+        // TODO: This timed out actually included two different messages:
+        // 1. Retries due to rate limiting: False.
+        // 2. Retries due to rate limiting: True.
+        // We might need to split this into two different rules after consult with DB team.
+        return error.message
+            && (error.message as string) === NoPrimaryInReplicasetRule.message;
+    }
+
+    shouldRetry: boolean = false;
+}
+
+// this handles the pool destroyed error from client side. Should relies on reconnect instead of retry?
+class PoolDestroyedRule implements IMongoExceptionRetryRule {
+    private static readonly message1 = "pool destroyed";
+    private static readonly message2 = "server instance pool was destroyed";
+    match(error: any): boolean {
+        return error.code === 16
+            && error.message
+            && ((error.message as string) === PoolDestroyedRule.message1
+                || (error.message as string) === PoolDestroyedRule.message2);
+    }
+
+    shouldRetry: boolean = false;
+}
+
+// this handles the case where the request payload is too large from server.
+class RequestSizeLargeRule implements IMongoExceptionRetryRule {
+    private static readonly errorMsgPrefix =
+        "Error=16, Details='Response status code does not indicate success: RequestEntityTooLarge (413)";
+    match(error: any): boolean {
+        return error.code === 16
+            && error.errmsg
+            && (error.errmsg as string).startsWith(RequestSizeLargeRule.errorMsgPrefix);
+    }
+
+    shouldRetry: boolean = false;
+}
+
+// This handles request timeout from server without additional rate limit info.
+class RequestTimedNoRateLimitInfo implements IMongoExceptionRetryRule {
+    private static readonly errmsg = "Request timed out.";
+    match(error: any): boolean {
+        return error.code === 50
+            && error.errmsg
+            && (error.errmsg as string) === RequestTimedNoRateLimitInfo.errmsg;
+    }
+
+    shouldRetry: boolean = true;
+}
+
+// This handles request timeout from server with http info.
+class RequestTimedOutWithHttpInfo implements IMongoExceptionRetryRule {
+    private static readonly errmsgPrefix =
+        "Error=50, Details='Response status code does not indicate success: RequestTimeout (408);";
+    match(error: any): boolean {
+        return error.code === 50
+            && error.errmsg
+            && (error.errmsg as string).startsWith(RequestTimedOutWithHttpInfo.errmsgPrefix);
+    }
+
+    shouldRetry: boolean = true;
+}
+
+// This handles request timeout from server with additional rate limit info.
+class RequestTimedOutWithRateLimitTrue implements IMongoExceptionRetryRule {
```

Comment:
@znewton @hedasilv 

Based on the issue we saw this morning for global db, when server replied with `"Request timed out. Retries due to rate limiting: True."`, it looks like something we should probably not retry, because retry will just add more loads. So I break the rule to `RequestTimedOutWithRateLimitTrue` and `RequestTimedOutWithRateLimitTrue`, and not retry due to rate limiting, but still retry when it is not due to rate limiting.

Please let me know if you have any concerns, thx